src/pubsub/ua_pubsub_reader.c

```diff
@@ -1240,22 +1284,37 @@ UA_Server_DataSetReader_process(UA_Server *server, UA_DataSetReader *dataSetRead
         return;
     }
 
+    UA_ReaderGroup *rg = UA_ReaderGroup_findRGbyId(server, dataSetReader->linkedReaderGroup);
     if(dataSetMsg->header.dataSetMessageType == UA_DATASETMESSAGE_DATAKEYFRAME) {
         if(dataSetMsg->header.fieldEncoding != UA_FIELDENCODING_RAWDATA) {
             size_t anzFields = dataSetMsg->data.keyFrameData.fieldCount;
             if(dataSetReader->config.dataSetMetaData.fieldsSize < anzFields) {
                 anzFields = dataSetReader->config.dataSetMetaData.fieldsSize;
             }
 
-            if(dataSetReader->subscribedDataSetTarget.targetVariablesSize < anzFields) {
-                anzFields = dataSetReader->subscribedDataSetTarget.targetVariablesSize;
+            if(dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariablesSize < anzFields) {
+                anzFields = dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariablesSize;
             }
 
             UA_StatusCode retVal = UA_STATUSCODE_GOOD;
+            if(rg->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
+                for(UA_UInt16 i = 0; i < anzFields; i++) {
+                    if(dataSetMsg->data.keyFrameData.dataSetFields[i].hasValue) {
+                        if(dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariables[i].attributeId == UA_ATTRIBUTEID_VALUE) {
+                            memcpy((**(dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariables[i].externalDataValueSource)).value.data,
+                                   dataSetMsg->data.keyFrameData.dataSetFields[i].value.data,
+                                   dataSetMsg->data.keyFrameData.dataSetFields[i].value.type->memSize);
```

Comment:
Here we are accessing the 'user memory'. The external data source PR allows the user to register callbacks to get informed if there is read or write access. Currently your implementation doesn't trigger this CB.